lib/getrange.c

```diff
@@ -58,7 +59,7 @@ getrange(const char *range,
 		if (!isdigit((unsigned char) *end))
 			return -1;
 
-		if (a2ul(max, end, NULL, 10, 0, ULONG_MAX) == -1)
+		if (a2ul(max, end, NULL, 10, *min, ULONG_MAX) == -1)
```

Comment:
Where is `ERANGE` reported in this code that you changed?

Reply:
a2ul(3) returns -1 and sets errno to ERANGE if the parsed value was out of range.  That's documented in
<https://www.alejandro-colomar.es/share/dist/liba2i/0/0.8/liba2i-0.8.pdf#pdf:bm13>.

The valid range is documented to be [min, max]:
<https://www.alejandro-colomar.es/share/dist/liba2i/0/0.8/liba2i-0.8.pdf#pdf:bm10>

where min and max are the last two parameters:
<https://www.alejandro-colomar.es/share/dist/liba2i/0/0.8/liba2i-0.8.pdf#pdf:bm9>

----

Or with code:

```c
$ grepc a2ul .
./lib/atoi/a2i.h:#define a2ul(n, s, ...)                                                       \
(                                                                             \
	_Generic(s,                                                           \
		const char *:  a2ul_c,                                        \
		const void *:  a2ul_c,                                        \
		char *:        a2ul_nc,                                       \
		void *:        a2ul_nc                                        \
	)(n, s, __VA_ARGS__)                                                  \
)
```

```c
$ grepc -tfd a2ul_c .
./lib/atoi/a2i.h:inline int
a2ul_c(unsigned long *restrict n, const char *s,
    const char **restrict endp, int base, unsigned long min, unsigned long max)
{
	return a2ul(n, (char *) s, (char **) endp, base, min, max);
}
```

(re-enters the macro from above, now with a non-const type)

```c
$ grepc -tfd a2ul_nc .
./lib/atoi/a2i.h:inline int
a2ul_nc(unsigned long *restrict n, char *s,
    char **restrict endp, int base, unsigned long min, unsigned long max)
{
	int  status;

	*n = strtou_noneg(s, endp, base, min, max, &status);
	if (status != 0) {
		errno = status;
		return -1;
	}
	return 0;
}
```

```c
$ grepc -tfd strtou_noneg .
./lib/atoi/strtou_noneg.h:inline uintmax_t
strtou_noneg(const char *s, char **restrict endp, int base,
    uintmax_t min, uintmax_t max, int *restrict status)
{
	int  st;

	if (status == NULL)
		status = &st;
	if (strtoi_(s, endp, base, 0, 1, status) == 0 && *status == ERANGE)
		return min;

	return strtou_(s, endp, base, min, max, status);
}
```

```c
$ grepc -tfd strtou_ .
./lib/atoi/strtoi.h:inline uintmax_t
strtou_(const char *s, char **restrict endp, int base,
    uintmax_t min, uintmax_t max, int *restrict status)
{
	return strtoN(s, endp, base, min, max, status, uintmax_t);
}
```

```c
$ grepc strtoN .
./lib/atoi/strtoi.h:#define strtoN(s, endp, base, min, max, status, TYPE)                         \
({                                                                            \
	const char  *s_ = s;                                                  \
	char        **endp_ = endp;                                           \
	int         base_ = base;                                             \
	TYPE        min_ = min;                                               \
	TYPE        max_ = max;                                               \
	int         *status_ = status;                                        \
                                                                              \
	int         e_, st_;                                                  \
	char        *end_;                                                    \
	TYPE        n_;                                                       \
                                                                              \
	if (endp_ == NULL)                                                    \
		endp_ = &end_;                                                \
	if (status_ == NULL)                                                  \
		status_ = &st_;                                               \
                                                                              \
	if (base_ != 0 && (base_ < 0 || base_ > 36)) {                        \
		*status_ = EINVAL;                                            \
		n_ = 0;                                                       \
                                                                              \
	} else {                                                              \
		e_ = errno;                                                   \
		errno = 0;                                                    \
		n_ = strtoNmax(TYPE, s_, endp_, base_);                       \
                                                                              \
		if (*endp_ == s_)                                             \
			*status_ = ECANCELED;                                 \
		else if (errno == ERANGE || n_ < min_ || n_ > max_)           \
			*status_ = ERANGE;                                    \
		else if (**endp_ != '\0')                                     \
			*status_ = ENOTSUP;                                   \
		else                                                          \
			*status_ = 0;                                         \
                                                                              \
		errno = e_;                                                   \
	}                                                                     \
	MAX(min_, MIN(max_, n_));                                             \
})
```

In the above macro, the interesting part is:

```c
		else if (errno == ERANGE || n_ < min_ || n_ > max_)           \
			*status_ = ERANGE;                                    \
```

Reply:
I understood that part, but I was wondering what `ERANGE` has to do with the code from this project. The commit message states: `Report an ERANGE error when min>max`, but I fail to see which part of shadow's code has anything to do with `ERANGE`.

Don't take me wrong, I think the code is fine, what I think is a little bit misleading is the commit message.

Reply:
Ahh, yeah, ERANGE doesn't have much to do with shadow's code.  I could have used `Report an error when min>max`.  I thought ERANGE would make it easier to read, but maybe it's the other way around.  I'll try to reword the commit.  :-)

Reply:
Done.